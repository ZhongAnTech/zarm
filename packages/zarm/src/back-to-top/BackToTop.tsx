import React, { useEffect } from 'react';
import { canUseDOM, ContainerType } from '../utils/dom';
import renderToContainer from '../utils/renderToContainer';
import Content, { BackTopTopProps } from './Content';

const BackToTop = (props: BackTopTopProps) => {
  const { mountContainer } = props;
  let unmount = () => {};

  useEffect(() => {
    unmount = renderToContainer(mountContainer as ContainerType, <Content {...props} />);
    return () => unmount();
  }, []);

  return null;
};

BackToTop.displayName = 'BackToTop';

BackToTop.defaultProps = {
  speed: 100,
  visibleDistance: 400,
  scrollContainer: canUseDOM ? window : undefined,
};

export default BackToTop;
