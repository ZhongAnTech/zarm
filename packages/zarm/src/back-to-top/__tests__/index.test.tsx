/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable prefer-destructuring */
/* eslint-disable dot-notation */
import React from 'react';
import type { CSSProperties, ReactElement } from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import BackToTop from '../index';
import Scroller from '../../scroller';

function genItems(count: number): ReactElement[] {
  const list: ReactElement[] = [];
  for (let i = 0; i < count; i++) {
    list.push(<li key={+i}>第 {i + 1} 行</li>);
  }
  return list;
}

describe('BackToTop', () => {
  let scrollTo: typeof window.scrollTo;
  let BackToTopCJS: typeof import('../index').default;
  beforeAll(() => {
    scrollTo = window.scrollTo;
    window.scrollTo = jest.fn();
  });
  beforeEach(() => {
    jest.resetModules();
  });
  afterEach(() => {
    jest.restoreAllMocks();
  });
  afterAll(() => {
    window.scrollTo = scrollTo;
  });
  describe('snapshot', () => {
    it('renders correctly', () => {
      const wrapper = mount(<BackToTop>Up</BackToTop>);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('scrollContainer', () => {
      const containerStyle: CSSProperties = {
        overflowY: 'auto',
        maxHeight: 400,
      };
      let container;
      const wrapper = mount(
        <>
          <ul
            ref={(ele) => {
              container = ele;
            }}
            style={containerStyle}
          >
            {genItems(100)}
          </ul>
          <BackToTop scrollContainer={() => container}>Up</BackToTop>
        </>,
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  it('should handle click event and scroll to the top with animation', () => {
    const onClick = jest.fn();
    const wrapper = shallow(<BackToTop onClick={onClick}>Up</BackToTop>);
    window.scrollTo(0, 1000);
    jest.useFakeTimers();
    wrapper.find('.za-back-to-top').simulate('click');
    jest.runAllTimers();
    expect(window.scrollTo).toBeCalledWith(0, 0);
    expect(onClick).toBeCalled();
  });

  it('should scroll to the top immediately without animation ', () => {
    const wrapper = shallow(<BackToTop speed={0}>Up</BackToTop>);
    window.scrollTo(0, 1000);
    wrapper.find('.za-back-to-top').simulate('click');
    expect(window.scrollTo).toBeCalledWith(0, 0);
  });

  it('should append portal container to document body when component did mount', () => {
    const appendChildSpy = jest.spyOn(document.body, 'appendChild');
    const createElementSpy = jest.spyOn(document, 'createElement');
    const wrapper = shallow(<BackToTop>Up</BackToTop>);
    expect(createElementSpy).toBeCalledWith('div');
    expect(appendChildSpy).toBeCalledWith(wrapper.instance()['portalContainer']);
    const portalContainer = document.body.querySelector('.za-back-to-top-container');
    expect(portalContainer).toBeTruthy();
  });

  it('should remove portal container from the document body when component will unmount', () => {
    const removeChildSpy = jest.spyOn(document.body, 'removeChild');
    const wrapper = shallow(<BackToTop>Up</BackToTop>);
    const portalContainer = wrapper.instance()['portalContainer'];
    wrapper.unmount();
    expect(removeChildSpy).toBeCalledWith(portalContainer);
  });

  it('should render null if environment does NOT support DOM', () => {
    jest.doMock('../../utils/dom', () => {
      const oDom = jest.requireActual('../../utils/dom');
      return {
        ...oDom,
        canUseDOM: false,
      };
    });
    BackToTopCJS = require('../index').default;
    const wrapper = shallow(<BackToTopCJS>Up</BackToTopCJS>);
    expect(BackToTopCJS.defaultProps.scrollContainer).toBeUndefined();
    expect(wrapper.instance()['parent']).toEqual(document.body);
    expect(wrapper.isEmptyRender()).toBeTruthy();
  });

  it('should handle scroll and set visible to false if scrollTop greater than props.visibleDistance', () => {
    const wrapper = shallow(<BackToTop>Up</BackToTop>);
    wrapper.find(Scroller).invoke('onScroll')!(300);
    expect(wrapper.find('.za-back-to-top').prop('style')).toEqual({
      display: 'none',
      position: 'fixed',
      bottom: 50,
      right: 50,
    });
  });

  it('should use scrollContainer as this.parent', () => {
    const scrollContainer = document.createElement('div');
    const wrapper = mount(<BackToTop scrollContainer={scrollContainer}>Up</BackToTop>);
    expect(wrapper.instance()['parent']).toEqual(scrollContainer);
  });

  it('should append new portal container if scrollContainer changed', () => {
    const appendChildSpy = jest.spyOn(document.body, 'appendChild');
    const scrollContainer = document.createElement('div');
    const wrapper = mount(<BackToTop>Up</BackToTop>);
    wrapper.setProps({ scrollContainer });
    expect(appendChildSpy).toBeCalledWith(wrapper.instance()['portalContainer']);
  });

  it('should render portal with correct style', () => {
    const scrollContainer = document.createElement('div');
    const wrapper = mount(
      <BackToTop scrollContainer={scrollContainer} style={{ color: 'blue' }}>
        Up
      </BackToTop>,
    );
    wrapper.find(Scroller).invoke('onScroll')!(500);
    expect(wrapper.find('.za-back-to-top').prop('style')).toEqual({
      display: 'inline',
      position: 'absolute',
      bottom: 50,
      right: 50,
      color: 'blue',
    });
  });
});
