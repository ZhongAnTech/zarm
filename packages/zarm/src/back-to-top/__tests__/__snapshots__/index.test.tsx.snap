// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackToTop snapshot renders correctly 1`] = `
<BackToTop
  prefixCls="za-back-to-top"
  scrollContainer={[Window]}
  speed={100}
  visibleDistance={400}
>
  <Portal
    containerInfo={
      <div
        class="za-back-to-top-container"
      >
        <div
          class="za-back-to-top"
          style="display: none; position: fixed; bottom: 50px; right: 50px;"
        >
          Up
        </div>
      </div>
    }
  >
    <div
      className="za-back-to-top"
      onClick={[Function]}
      style={
        Object {
          "bottom": 50,
          "display": "none",
          "position": "fixed",
          "right": 50,
        }
      }
    >
      Up
    </div>
    <Scroller
      container={[Window]}
      onScroll={[Function]}
      prefixCls="za-scroller"
    />
  </Portal>
</BackToTop>
`;

exports[`BackToTop snapshot scrollContainer 1`] = `
Array [
  <ul
    style={
      Object {
        "maxHeight": 400,
        "overflowY": "auto",
      }
    }
  >
    <li
      key="0"
    >
      第 
      1
       行
    </li>
    <li
      key="1"
    >
      第 
      2
       行
    </li>
    <li
      key="2"
    >
      第 
      3
       行
    </li>
    <li
      key="3"
    >
      第 
      4
       行
    </li>
    <li
      key="4"
    >
      第 
      5
       行
    </li>
    <li
      key="5"
    >
      第 
      6
       行
    </li>
    <li
      key="6"
    >
      第 
      7
       行
    </li>
    <li
      key="7"
    >
      第 
      8
       行
    </li>
    <li
      key="8"
    >
      第 
      9
       行
    </li>
    <li
      key="9"
    >
      第 
      10
       行
    </li>
    <li
      key="10"
    >
      第 
      11
       行
    </li>
    <li
      key="11"
    >
      第 
      12
       行
    </li>
    <li
      key="12"
    >
      第 
      13
       行
    </li>
    <li
      key="13"
    >
      第 
      14
       行
    </li>
    <li
      key="14"
    >
      第 
      15
       行
    </li>
    <li
      key="15"
    >
      第 
      16
       行
    </li>
    <li
      key="16"
    >
      第 
      17
       行
    </li>
    <li
      key="17"
    >
      第 
      18
       行
    </li>
    <li
      key="18"
    >
      第 
      19
       行
    </li>
    <li
      key="19"
    >
      第 
      20
       行
    </li>
    <li
      key="20"
    >
      第 
      21
       行
    </li>
    <li
      key="21"
    >
      第 
      22
       行
    </li>
    <li
      key="22"
    >
      第 
      23
       行
    </li>
    <li
      key="23"
    >
      第 
      24
       行
    </li>
    <li
      key="24"
    >
      第 
      25
       行
    </li>
    <li
      key="25"
    >
      第 
      26
       行
    </li>
    <li
      key="26"
    >
      第 
      27
       行
    </li>
    <li
      key="27"
    >
      第 
      28
       行
    </li>
    <li
      key="28"
    >
      第 
      29
       行
    </li>
    <li
      key="29"
    >
      第 
      30
       行
    </li>
    <li
      key="30"
    >
      第 
      31
       行
    </li>
    <li
      key="31"
    >
      第 
      32
       行
    </li>
    <li
      key="32"
    >
      第 
      33
       行
    </li>
    <li
      key="33"
    >
      第 
      34
       行
    </li>
    <li
      key="34"
    >
      第 
      35
       行
    </li>
    <li
      key="35"
    >
      第 
      36
       行
    </li>
    <li
      key="36"
    >
      第 
      37
       行
    </li>
    <li
      key="37"
    >
      第 
      38
       行
    </li>
    <li
      key="38"
    >
      第 
      39
       行
    </li>
    <li
      key="39"
    >
      第 
      40
       行
    </li>
    <li
      key="40"
    >
      第 
      41
       行
    </li>
    <li
      key="41"
    >
      第 
      42
       行
    </li>
    <li
      key="42"
    >
      第 
      43
       行
    </li>
    <li
      key="43"
    >
      第 
      44
       行
    </li>
    <li
      key="44"
    >
      第 
      45
       行
    </li>
    <li
      key="45"
    >
      第 
      46
       行
    </li>
    <li
      key="46"
    >
      第 
      47
       行
    </li>
    <li
      key="47"
    >
      第 
      48
       行
    </li>
    <li
      key="48"
    >
      第 
      49
       行
    </li>
    <li
      key="49"
    >
      第 
      50
       行
    </li>
    <li
      key="50"
    >
      第 
      51
       行
    </li>
    <li
      key="51"
    >
      第 
      52
       行
    </li>
    <li
      key="52"
    >
      第 
      53
       行
    </li>
    <li
      key="53"
    >
      第 
      54
       行
    </li>
    <li
      key="54"
    >
      第 
      55
       行
    </li>
    <li
      key="55"
    >
      第 
      56
       行
    </li>
    <li
      key="56"
    >
      第 
      57
       行
    </li>
    <li
      key="57"
    >
      第 
      58
       行
    </li>
    <li
      key="58"
    >
      第 
      59
       行
    </li>
    <li
      key="59"
    >
      第 
      60
       行
    </li>
    <li
      key="60"
    >
      第 
      61
       行
    </li>
    <li
      key="61"
    >
      第 
      62
       行
    </li>
    <li
      key="62"
    >
      第 
      63
       行
    </li>
    <li
      key="63"
    >
      第 
      64
       行
    </li>
    <li
      key="64"
    >
      第 
      65
       行
    </li>
    <li
      key="65"
    >
      第 
      66
       行
    </li>
    <li
      key="66"
    >
      第 
      67
       行
    </li>
    <li
      key="67"
    >
      第 
      68
       行
    </li>
    <li
      key="68"
    >
      第 
      69
       行
    </li>
    <li
      key="69"
    >
      第 
      70
       行
    </li>
    <li
      key="70"
    >
      第 
      71
       行
    </li>
    <li
      key="71"
    >
      第 
      72
       行
    </li>
    <li
      key="72"
    >
      第 
      73
       行
    </li>
    <li
      key="73"
    >
      第 
      74
       行
    </li>
    <li
      key="74"
    >
      第 
      75
       行
    </li>
    <li
      key="75"
    >
      第 
      76
       行
    </li>
    <li
      key="76"
    >
      第 
      77
       行
    </li>
    <li
      key="77"
    >
      第 
      78
       行
    </li>
    <li
      key="78"
    >
      第 
      79
       行
    </li>
    <li
      key="79"
    >
      第 
      80
       行
    </li>
    <li
      key="80"
    >
      第 
      81
       行
    </li>
    <li
      key="81"
    >
      第 
      82
       行
    </li>
    <li
      key="82"
    >
      第 
      83
       行
    </li>
    <li
      key="83"
    >
      第 
      84
       行
    </li>
    <li
      key="84"
    >
      第 
      85
       行
    </li>
    <li
      key="85"
    >
      第 
      86
       行
    </li>
    <li
      key="86"
    >
      第 
      87
       行
    </li>
    <li
      key="87"
    >
      第 
      88
       行
    </li>
    <li
      key="88"
    >
      第 
      89
       行
    </li>
    <li
      key="89"
    >
      第 
      90
       行
    </li>
    <li
      key="90"
    >
      第 
      91
       行
    </li>
    <li
      key="91"
    >
      第 
      92
       行
    </li>
    <li
      key="92"
    >
      第 
      93
       行
    </li>
    <li
      key="93"
    >
      第 
      94
       行
    </li>
    <li
      key="94"
    >
      第 
      95
       行
    </li>
    <li
      key="95"
    >
      第 
      96
       行
    </li>
    <li
      key="96"
    >
      第 
      97
       行
    </li>
    <li
      key="97"
    >
      第 
      98
       行
    </li>
    <li
      key="98"
    >
      第 
      99
       行
    </li>
    <li
      key="99"
    >
      第 
      100
       行
    </li>
  </ul>,
  <BackToTop
    prefixCls="za-back-to-top"
    scrollContainer={[Function]}
    speed={100}
    visibleDistance={400}
  >
    <Portal
      containerInfo={
        <div
          class="za-back-to-top-container"
        >
          <div
            class="za-back-to-top"
            style="display: none; position: fixed; bottom: 50px; right: 50px;"
          >
            Up
          </div>
        </div>
      }
    >
      <div
        className="za-back-to-top"
        onClick={[Function]}
        style={
          Object {
            "bottom": 50,
            "display": "none",
            "position": "fixed",
            "right": 50,
          }
        }
      >
        Up
      </div>
      <Scroller
        container={[Function]}
        onScroll={[Function]}
        prefixCls="za-scroller"
      />
    </Portal>
  </BackToTop>,
]
`;
